### PR TITLE
feat: add PHP 8 support

### LINKED ISSUE #5: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by alexraputa**
Closes #5 

|    Q          |   A
|-------------- | ------
| New Feature   | yes

### CODE REVIEW

**Comment by alexraputa** [comment]
@boesing, ready for review.

File: composer.json [hunk 1]
```diff
@@ -29,15 +29,17 @@
         }
     },
     "require": {
-        "php": "^7.1",
+        "php": "^7.3 || ^8.0",
```
Comment (froschdesign) [inline composer.json]: Please recheck to list in [related issue report](https://github.com/mezzio/mezzio-authentication/issues/5):

> Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`

Thanks in advance!

Comment (alexraputa) [reply composer.json]: @froschdesign, fixed.

File: phpunit.xml.dist [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
+>
+    <coverage processUncoveredFiles="true">
+        <include>
+            <directory suffix=".php">src</directory>
+        </include>
+    </coverage>
+
     <testsuites>
-        <testsuite name="mezzio-authentication">
+        <testsuite name="mezzio-authentication unit tests">
```
Comment (geerteltink) [inline phpunit.xml.dist]: You can leave the testsuite name as it is.

File: test/AuthenticationMiddlewareFactoryTest.php [hunk 3]
```diff
@@ -13,27 +13,30 @@
...
 
 class AuthenticationMiddlewareFactoryTest extends TestCase
 {
+    use ProphecyTrait;
+
     protected $authentication;
     protected $request;
 
-    public function setUp()
+    public function setUp(): void
```
Comment (geerteltink) [inline test/AuthenticationMiddlewareFactoryTest.php]:
```suggestion
    protected function setUp(): void
```

File: test/AuthenticationMiddlewareTest.php [hunk 4]
```diff
@@ -14,17 +14,20 @@
...
 
 class AuthenticationMiddlewareTest extends TestCase
 {
+    use ProphecyTrait;
+
     protected $authentication;
     protected $request;
 
-    public function setUp()
+    public function setUp(): void
```
Comment (geerteltink) [inline test/AuthenticationMiddlewareTest.php]:
```suggestion
    protected function setUp(): void
```

File: test/DefaultUserFactoryTest.php [hunk 5]
```diff
@@ -14,11 +14,14 @@
...
 use PHPUnit\Framework\TestCase;
+use Prophecy\PhpUnit\ProphecyTrait;
 use Psr\Container\ContainerInterface;
 
 class DefaultUserFactoryTest extends TestCase
 {
-    public function setUp()
+    use ProphecyTrait;
+
+    public function setUp(): void
```
Comment (geerteltink) [inline test/DefaultUserFactoryTest.php]:
```suggestion
    protected function setUp(): void
```

**Review by geerteltink — CHANGES_REQUESTED** [review]

File: phpunit.xml.dist (hunk 2 above)
Comment (alexraputa) [reply phpunit.xml.dist]: > You can leave the testsuite name as it is.

Fixed.

File: composer.json [hunk 6]
```diff
@@ -29,15 +29,17 @@
...
         "psr/container": "^1.0",
         "psr/http-message": "^1.0.1",
         "psr/http-server-middleware": "^1.0"
     },
     "require-dev": {
         "laminas/laminas-coding-standard": "~1.0.0",
-        "phpunit/phpunit": "^7.0.2"
+        "phpspec/prophecy": "^1.12",
+        "phpspec/prophecy-phpunit": "^2.0",
+        "phpunit/phpunit": "^9.4"
```
Comment (geerteltink) [inline composer.json]: As mentioned in #5 it should be phpunit ^9.3.

File: phpunit.xml.dist [hunk 7]
```diff
@@ -1,18 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
-         bootstrap="vendor/autoload.php"
-         colors="true"
-         stderr="true">
+<phpunit
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:noNamespaceSchemaLocation="vendor/phpunit/phpunit/phpunit.xsd"
```
Comment (geerteltink) [inline phpunit.xml.dist]: When running phpunit you probably get a warning about an invalid schema. When you let phpunit fix it, you get something like this 
```
xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd" 
```

File: composer.json (hunk 6 above)
Comment (alexraputa) [reply composer.json]: Done.

File: phpunit.xml.dist (hunk 7 above)
Comment (alexraputa) [reply phpunit.xml.dist]: Done.

**Review by geerteltink — APPROVED** [review]

File: src/UserRepository/PdoDatabaseFactory.php [hunk 8]
```diff
@@ -59,12 +59,11 @@ public function __invoke(ContainerInterface $container) : PdoDatabase
             );
         }
 
+        $connection = new PDO($pdo['dsn'], $pdo['username'] ?? null, $pdo['password'] ?? null);
+        $connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_SILENT);
```
Comment (froschdesign) [inline src/UserRepository/PdoDatabaseFactory.php]: I think this should be removed in a future version therefore a new issue report should be opened for this.

Reference:

* https://github.com/php/php-src/blob/ae944503c34d7c89433eaefe607fc0d631d6f9ac/UPGRADING#L424-L428
* https://wiki.php.net/rfc/pdo_default_errmode

File: src/AuthenticationMiddleware.php [hunk 9]
```diff
@@ -22,9 +22,6 @@ class AuthenticationMiddleware implements MiddlewareInterface
      */
     protected $auth;
 
-    /**
-     * @param AuthenticationInterface $auth
-     */
```
Comment (alexraputa) [inline src/AuthenticationMiddleware.php:27]: Redundant PHPDoc.

**Comment by alexraputa** [comment]
@boesing, can you review this PR? 🙏 ;)

File: src/UserRepository/PdoDatabaseFactory.php [hunk 10]
```diff
@@ -72,14 +72,13 @@ public function __invoke(ContainerInterface $container) : PdoDatabase
...
-                (string) ($pdo['password'] ?? null)
-            ),
-            $pdo,
-            $user
+        $connection = new PDO(
+            (string) $pdo['dsn'],
+            (string) ($pdo['username'] ?? null),
+            (string) ($pdo['password'] ?? null)
         );
+        $connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_SILENT);
```
Comment (boesing) [inline src/UserRepository/PdoDatabaseFactory.php:81]: We might add `@deprecated` here to enable maintainers to search for deprecations in next versions.
I guess thats what was requested from @froschdesign in https://github.com/mezzio/mezzio-authentication/pull/7#discussion_r504876982

**Review by boesing — CHANGES_REQUESTED** [review]
LGTM. I've just added one change request to address https://github.com/mezzio/mezzio-authentication/pull/7#discussion_r504876982

File: src/UserRepository/PdoDatabaseFactory.php (hunk 10 above)
Comment (alexraputa) [reply src/UserRepository/PdoDatabaseFactory.php:81]: Done.

**Review by boesing — APPROVED** [review]

**Comment by boesing** [comment]
After thinking about the fact that we are marking the error mode deprecated, we might consider adding options in the future to configure the error mode. Thus, I wont mention this deprecation in the release notes as I think we might add a configuration entry for that error mode in the future.

**Comment by boesing** [comment]
Thanks, @alexraputa

**Pull request merged**